panic: accept routine exited: filter timed out

```
panic: accept routine exited: filter timed out

goroutine 791038 [running]:
github.com/tendermint/tendermint/p2p.(*Switch).acceptRoutine(0xc000352690)
        /home/jslave/workspace/02.Fetch.Binary@3/go/src/github.com/tendermint/tendermint/p2p/switch.go:516 +0xead
created by github.com/tendermint/tendermint/p2p.(*Switch).OnStart
        /home/jslave/workspace/02.Fetch.Binary@3/go/src/github.com/tendermint/tendermint/p2p/switch.go:204 +0x1ea
```

happend on tendermint 0.30.1 under high LA (50-100), machine was struggling with the load.

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -44,3 +44,4 @@ Special thanks to external contributors on this release:
 
 - [p2p/conn] \#3347 Reject all-zero shared secrets in the Diffie-Hellman step of secret-connection
 - [libs/pubsub] \#951, \#1880 use non-blocking send when dispatching messages [ADR-33](https://github.com/tendermint/tendermint/blob/develop/docs/architecture/adr-033-pubsub.md)
+- [p2p] \#3369 do not panic when filter times out

--- a/p2p/switch.go
+++ b/p2p/switch.go
@@ -504,7 +504,7 @@ func (sw *Switch) acceptRoutine() {
 				)
 
 				continue
-			case *ErrTransportClosed:
+			case ErrTransportClosed:
 				sw.Logger.Error(
 					"Stopped accept routine, as transport is closed",
 					"numPeers", sw.peers.Size(),

--- a/p2p/switch.go
+++ b/p2p/switch.go
@@ -497,6 +497,13 @@ func (sw *Switch) acceptRoutine() {
 				)
 
 				continue
+			case ErrFilterTimeout:
+				sw.Logger.Error(
+					"Peer filter timed out",
+					"err", err,
+				)
+
+				continue
 			case *ErrTransportClosed:
 				sw.Logger.Error(
 					"Stopped accept routine, as transport is closed",

--- a/p2p/switch_test.go
+++ b/p2p/switch_test.go
@@ -2,6 +2,7 @@ package p2p
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
@@ -530,6 +531,44 @@ func TestSwitchAcceptRoutine(t *testing.T) {
 	for _, rp := range remotePeers {
 		rp.Stop()
 	}
+}
+
+type errorTransport struct {
+	acceptErr error
+}
+
+func (et errorTransport) Accept(c peerConfig) (Peer, error) {
+	return nil, et.acceptErr
+}
+func (errorTransport) Dial(NetAddress, peerConfig) (Peer, error) {
+	panic("not implemented")
+}
+func (errorTransport) Cleanup(Peer) {
+	panic("not implemented")
+}
+
+func TestSwitchAcceptRoutineErrorCases(t *testing.T) {
+	sw := NewSwitch(cfg, errorTransport{ErrFilterTimeout{}})
+	assert.NotPanics(t, func() {
+		err := sw.Start()
+		assert.NoError(t, err)
+		sw.Stop()
+	})
+
+	sw = NewSwitch(cfg, errorTransport{ErrRejected{conn: nil, err: errors.New("filtered"), isFiltered: true}})
+	assert.NotPanics(t, func() {
+		err := sw.Start()
+		assert.NoError(t, err)
+		sw.Stop()
+	})
+	// TODO(melekes) check we remove our address from addrBook
+
+	sw = NewSwitch(cfg, errorTransport{ErrTransportClosed{}})
+	assert.NotPanics(t, func() {
+		err := sw.Start()
+		assert.NoError(t, err)
+		sw.Stop()
+	})
 }
 
 func BenchmarkSwitchBroadcast(b *testing.B) {

--- a/p2p/transport.go
+++ b/p2p/transport.go
@@ -175,7 +175,7 @@ func (mt *MultiplexTransport) Accept(cfg peerConfig) (Peer, error) {
 
 		return mt.wrapPeer(a.conn, a.nodeInfo, cfg, nil), nil
 	case <-mt.closec:
-		return nil, &ErrTransportClosed{}
+		return nil, ErrTransportClosed{}
 	}
 }
 

